### PR TITLE
test(amf): UT for the service request procedure

### PR DESCRIPTION
**Pull request by VinayKumar1591**
test(amf): UT for the service request procedure

## Summary

1. Added Unit test cases for the service request procedures
## Test Plan

1. Tested uisng Gtest

### CODE REVIEW

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 1]
```diff
@@ -8,19 +8,10 @@
...
-#include "lte/gateway/c/core/oai/lib/itti/intertask_interface.h"
-#include "../../tasks/amf/amf_app_ue_context_and_proc.h"
-#include "lte/gateway/c/core/oai/include/mme_config.h"
-#include "lte/gateway/c/core/oai/tasks/amf/amf_authentication.h"
-#include "lte/gateway/c/core/oai/test/amf/util_s6a_update_location.h"
-#include "lte/gateway/c/core/oai/tasks/amf/amf_recv.h"
-#include "lte/gateway/c/core/oai/tasks/amf/amf_identity.h"
-#include "lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.h"
-#include "lte/gateway/c/core/oai/tasks/amf/include/amf_smf_packet_handler.h"
+#include "../mock_tasks/mock_tasks.h"
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: include full path as in the other header files.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 2]
```diff
@@ -30,21 +21,45 @@ extern "C" {
...
 #include "lte/gateway/c/core/oai/common/itti_free_defined_msg.h"
+#include "lte/gateway/c/core/oai/common/common_types.h"
+#include "lte/gateway/c/core/oai/include/amf_config.h"
 }
 
+#include "lte/gateway/c/core/oai/test/amf/util_nas5g_pkt.h"
+#include "lte/gateway/c/core/oai/tasks/amf/include/amf_session_manager_pco.h"
+#include <gtest/gtest.h>
+#include "lte/gateway/c/core/oai/lib/itti/intertask_interface.h"
+#include "../../tasks/amf/amf_app_ue_context_and_proc.h"
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: include full path as in the other cases.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 3]
```diff
@@ -30,21 +21,45 @@ extern "C" {
...
+#include "lte/gateway/c/core/oai/include/mme_config.h"
+#include "lte/gateway/c/core/oai/tasks/amf/amf_authentication.h"
+#include "util_s6a_update_location.h"
+#include "tasks/amf/amf_recv.h"
+#include "tasks/amf/amf_identity.h"
+#include "tasks/amf/amf_app_ue_context_and_proc.h"
+#include "tasks/amf/amf_sap.h"
+#include "tasks/amf/amf_app_state_manager.h"
+#include "tasks/amf/amf_as.h"
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: include full path as in the other cases.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 4]
```diff
@@ -30,21 +21,45 @@ extern "C" {
...
 const task_info_t tasks_info[] = {
     {THREAD_NULL, "TASK_UNKNOWN", "ipc://IPC_TASK_UNKNOWN"},
 #define TASK_DEF(tHREADiD)                                                     \
   {THREAD_##tHREADiD, #tHREADiD, "ipc://IPC_" #tHREADiD},
 #include "lte/gateway/c/core/oai/include/tasks_def.h"
 #undef TASK_DEF
 };
-
-task_zmq_ctx_t grpc_service_task_zmq_ctx;
+#endif
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: remove if not used.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 5]
```diff
@@ -880,6 +907,372 @@ TEST(test_dnn, test_amf_validate_dnn) {
...
+#define AMF_TAC 0x03
+
+  ue_m5gmm_context_s* ue_context;
+  amf_app_desc_t* amf_app_desc_p;
+  guti_m5_t guti;
+  tai_t tai;
+  const amf_ue_ngap_id_t AMF_UE_NGAP_ID = 0x05;
+  const gnb_ue_ngap_id_t gNB_UE_NGAP_ID = 0x09;
+  const uint32_t gnb_id                 = 0x01;
+  const imsi64_t imis64                 = IMSI64;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: imis64 -> imsi64

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: IMSI64 looks redundant, why not directly set imsi64 to `222456000000101`

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 6]
```diff
@@ -880,6 +907,372 @@ TEST(test_dnn, test_amf_validate_dnn) {
...
+#define MCC_DIGIT3 2
+#define MNC_DIGIT1 4
+#define MNC_DIGIT2 5
+#define MNC_DIGIT3 6
+#define IMSI64 222456000000101
+#define M_TMSI 0X212e5025
+#define AMF_SET_ID 1
+#define AMF_POINTER 0
+#define AMF_REGION_ID 1
+#define AMF_TAC 0x03
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:910]: looks redundant to me. are they used anywhere else other than initializing the class variables?

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 7]
```diff
@@ -880,6 +907,372 @@ TEST(test_dnn, test_amf_validate_dnn) {
...
+    // AMF GUAMI CONFIG
+    amf_config.guamfi.guamfi[0].plmn.mcc_digit1 = MCC_DIGIT1;
+    amf_config.guamfi.guamfi[0].plmn.mcc_digit2 = MCC_DIGIT2;
+    amf_config.guamfi.guamfi[0].plmn.mcc_digit3 = MCC_DIGIT3;
+    amf_config.guamfi.guamfi[0].plmn.mnc_digit1 = MNC_DIGIT1;
+    amf_config.guamfi.guamfi[0].plmn.mnc_digit2 = MNC_DIGIT2;
+    amf_config.guamfi.guamfi[0].plmn.mnc_digit3 = MNC_DIGIT3;
+    amf_config.guamfi.guamfi[0].amf_set_id      = AMF_SET_ID;
+    amf_config.guamfi.guamfi[0].amf_pointer     = AMF_POINTER;
+    amf_config.guamfi.guamfi[0].amf_regionid    = AMF_REGION_ID;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:957]: move to a utility function to set default configs.

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 8]
```diff
@@ -880,6 +907,372 @@ TEST(test_dnn, test_amf_validate_dnn) {
...
+    // imsi64
+    ue_context->amf_context.imsi64 = imis64;
+    // ue security context
+    ue_context->amf_context.member_present_mask |= AMF_CTXT_MEMBER_SECURITY;
+    // ueContextReq
+    ue_context->ue_context_request = M5G_UEContextRequest_requested;
+    // ue state
+    ue_context->mm_state = REGISTERED_IDLE;
+    // 5G TMSI
+    ue_context->amf_context.m5_guti.m_tmsi = m_imsi;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: move to a utility function

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp (hunk 6 above)
Comment (VinayKumar1591) [reply lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:910]: used in testcases and will use in upcoming unit testcases too

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp (hunk 8 above)
Comment (VinayKumar1591) [reply lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: default config already part of setup method which is to be used for setting configs , initialization etc..

File: lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp [hunk 9]
```diff
@@ -99,12 +104,24 @@ uint8_t NAS5GPktSnapShot::deregistrarion_request[17] = {
...
 
+// service request with service type data
 uint8_t NAS5GPktSnapShot::service_request[37] = {
     0x7e, 0x00, 0x4c, 0x10, 0x00, 0x07, 0xf4, 0x00, 0x00, 0xe4,
     0x2c, 0x6c, 0x68, 0x71, 0x00, 0x15, 0x7e, 0x00, 0x4c, 0x10,
     0x00, 0x07, 0xf4, 0x00, 0x00, 0xe4, 0x2c, 0x6c, 0x68, 0x40,
     0x02, 0x20, 0x00, 0x50, 0x02, 0x20, 0x00};
 
+// service request with service type signaling
+uint8_t NAS5GPktSnapShot::service_req_signalling[13] = {
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp]: signalling -> signaling

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**